packages/slate-react/src/components/editor.js

```diff
@@ -228,17 +189,61 @@ class Editor extends React.Component {
    */
 
   get schema() {
-    return this.state.schema
+    const plugins = this.resolvePlugins(this.props.plugins, this.props.schema)
+    return this.getSchemaWithMemoization(plugins)
   }
 
   get stack() {
-    return this.state.stack
+    const plugins = this.resolvePlugins(this.props.plugins, this.props.schema)
+    return this.getStackWithMemoization(plugins)
   }
 
   get value() {
-    return this.state.value
+    return this.processValueOnChange(this.props.value, this.stack)
   }
 
+  /**
+   * Get stack by plugins, it is a memoized function and therefore do not need
+   * to re-evaluate for the same plugins
+   *
+   * @param {Array<Object>} plugins
+   * @return {Stack}
+   */
+
+  getStackWithMemoization = memoize(plugins => {
+    this.tmp.resolves++
+    return Stack.create({ plugins })
+  })
+
+  /**
+   * Get schema by plugins, it is a memoized function and therefore do not need
+   * to re-evaluate for the same plugins
+   *
+   * @param {Array<Object>} plugins
+   * @return {Schema}
+   */
+
+  getSchemaWithMemoization = memoize(plugins => Schema.create({ plugins }))
+
+  /**
+   * Get value after onChange stack, it is a memoized function and therefore do not need
+   * to re-evaluate for the same value and stack.
+   *
+   * If the value is produced by onChange with the same stack, then will immediately return
+   * this value.
+   *
+   * @param {Value} value
+   * @return {Schema}
+   */
+
+  processValueOnChange = memoize((value, stack) => {
+    if (this.tmp.change && value === this.tmp.change.value) return value
```

Comment:
if we are going to memoize this, should we explicitly pass in `this.tmp.change.value` as an argument?

or are there other protections against `this.tmp.change.value` changing between calls to this method while `value` and `stack` stay the same?

Reply:
This part is to ensure  `processValueOnChange(processValueOnChange(value)) === processValueOnChange(value)`

The reason is that, in each update, we will have `stack.run('onChange', change)` passed to the parent container, then the parent container will pass `stack.run('onChange', change)` back as `props.value`.  And I do not want to have the `stack.run('onChange', change)` run again when receiving props.

I do not see a neat way to solve the problem.  The only thing I can do is to ensure all value are updates are passed via `processValueOnChange`

Reply:
@jtadmor Hi, I have just removed this part of logic in recent commit.   I agree with you that we shall leave the memoized function pure.

Cheers,